rockcraft/plugins/maven_plugin.py

```diff
@@ -0,0 +1,76 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2024 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""The Rockcraft Maven plugin."""
+import logging
+from textwrap import dedent
+
+from craft_parts.plugins import maven_plugin
+from overrides import override  # type: ignore[reportUnknownVariableType]
+
+logger = logging.getLogger(__name__)
+
+# Template for the sitecustomize module that we'll add to the payload so that
+# the pip-installed packages are found regardless of how the interpreter is
+# called.
+SITECUSTOMIZE_TEMPLATE = dedent(
+    """
+    # sitecustomize added by Rockcraft.
+    import site
+    import sys
+
+    major, minor = sys.version_info.major, sys.version_info.minor
+    site_dir = f"/lib/python{major}.{minor}/site-packages"
+    dist_dir = "/usr/lib/python3/dist-packages"
+
+    # Add the directory that contains the venv-installed packages.
+    site.addsitedir(site_dir)
+
+    if dist_dir in sys.path:
+        # Make sure that this site-packages dir comes *before* the base-provided
+        # dist-packages dir in sys.path.
+        path = sys.path
+        site_index = path.index(site_dir)
+        dist_index = path.index(dist_dir)
+
+        if dist_index < site_index:
+            path[dist_index], path[site_index] = path[site_index], path[dist_index]
+
+    EOF
+    """
+).strip()
+
+
+class MavenPlugin(maven_plugin.MavenPlugin):
+    """A MavenPlugin plugin for Rockcraft.
+
+    This plugin extends Craft-parts' vanilla Maven plugin to properly
+    use and install the Java VM. Specifically:
+
+    - Do not link java executable to /bin/java
+    - Do not include staging area in the PATH
+    """
+
+    @override
+    def _get_java_link_commands(self) -> list[str]:
+        """Get the bash commands to provide /bin/java symlink."""
+        return []
+
+    @override
+    def get_build_environment(self) -> dict[str, str]:
+        env = super().get_build_environment()
+        env["PATH"] = "/usr/bin"
```

Comment:
This is a pretty big change that we should only do if there's absolutely no other way around it. The `PATH` that craft-parts builds prioritizes the install and stage dirs on purpose, to let you do things like using a custom jdk in your build.

This is needed because we need Maven to pick the build-package jdk instead of the staged jre, right? Could we configure `JAVA_HOME` somehow? Like, to `"$(dirname $(dirname $(readlink -f $(which javac))))"` (the directory that contains `bin/javac`)?

Reply:
I think there are two different problems:
 - maven project does not build because selected toolchain is wrong. Auto detected java home solves this.  
 - part build is not reproducible and the build output is not what I expected as a user, because the the build pick up different toolchains. The edge case - project fails to build at all. Limiting path to /usr/bin solves this, but not in line with other part plugins environment settings.   Another approach would be to sandbox staging directories to each part and ensure that it can not access output of its own staging phase.

Reply:
Maybe as a bandaid just put /usr/bin in front of the PATH?

Reply:
Maybe as a bandaid just put /usr/bin in front of the PATH?